### PR TITLE
OSX use fontAttributes for better emulation of bold fonts

### PR DESCRIPTION
**Pull request by csomor**
Using fontAttributes dictionary allows for better emulation of bold typefaces, also switch to using color from context, this allows keeping the same dict throughout

Thanks in advance for reviewing this I hope I didn't overlook something ...

### CODE REVIEW

File: src/osx/carbon/font.cpp [hunk 1]
```diff
@@ -366,6 +393,11 @@ void wxFontRefData::MacFindFont()
                         m_ctFont.reset(fontWithTraits);
                 }
             }
+            CFDictionarySetValue(dict, kCTFontAttributeName, m_ctFont.get() );
+            CFDictionarySetValue(dict, kCTForegroundColorFromContextAttributeName, kCFBooleanTrue);
+            
+            fontcache[ std::wstring(lookupnameWithSize.wc_str()) ] = m_ctFont;
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: `s.ToStdWstring()` is a shorter and more clear way of writing `std::wstring(s.wc_str())` (this applies here and in a few other places).

File: src/osx/carbon/graphics.cpp [hunk 2]
```diff
@@ -2333,10 +2322,9 @@ void wxMacCoreGraphicsContext::GetTextExtent( const wxString &str, wxDouble *wid
     {
         if ( width )
             *width = w;
-        if ( height )
-            *height = a+d+l;
     }
-
+    if ( height )
```
Comment (vadz) [inline src/osx/carbon/graphics.cpp]: Not sure why has this been moved? It seems logical to return height of 0 for an empty string and I suspect that the other ports (still) do the same. If we do need to change it, it would be better to do it in a separate commit and, ideally, with a unit test checking for this, so that we could ensure the behaviour is consistent across all ports.

File: src/osx/carbon/font.cpp [hunk 3]
```diff
@@ -307,9 +314,18 @@ void wxFontRefData::MacFindFont()
         wxString lookupnameWithSize = wxString::Format( "%s_%u_%d", m_info.m_faceName, traits, m_info.m_pointSize );
 
         static std::map< std::wstring , wxCFRef< CTFontRef > > fontcache ;
+        static std::map< std::wstring , wxCFRef< CFDictionaryRef > > fontattributescache ;
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: Looking at how this map is used, it seems it would be better to have a single map of font name+size to pair/struct of font and attributes refs. I.e. why not change `fontcache` value type instead of adding another map?

File: src/osx/carbon/font.cpp [hunk 4]
```diff
@@ -350,11 +366,22 @@ void wxFontRefData::MacFindFont()
...
+                                {
+                                    remainingTraits &= ~kCTFontBoldTrait;
+                                }
                             }
                         }
+                        
+                        // we have to emulate bold
+                        if ( remainingTraits & kCTFontBoldTrait )
+                        {
+                            const float strokewidth = -3.0;
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: It would be nice to have a comment saying where does this mysterious (to me, at least) `-3` come from.

File: src/osx/carbon/font.cpp [hunk 5]
```diff
@@ -659,6 +691,16 @@ CTFontRef wxFont::OSXGetCTFont() const
     return (CTFontRef)(M_FONTDATA->m_ctFont);
 }
 
+CFDictionaryRef wxFont::OSXGetCTFontAttributes() const
+{
+    wxCHECK_MSG( M_FONTDATA != NULL , 0, wxT("invalid font") );
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: Very minor, but this `0` should be `NULL` here, for consistency -- we use `NULL`s for other `XXXRef`s elsewhere.

**Review by vadz — APPROVED** [review]
No real comments except, perhaps, for the height change.

Thanks for improving this (does it merit an entry in the change log, BTW?)!

File: src/osx/carbon/graphics.cpp (hunk 2 above)
Comment (csomor) [inline src/osx/carbon/graphics.cpp]: oh, sorry, I thought I had made a mistake when writing it. I had expected it to still return the height of the font rectangle and stumbled upon it when using it, it seemed so obvious to be a bug on my part that I fixed without even checking the other port

File: src/osx/carbon/font.cpp (hunk 4 above)
Comment (csomor) [inline src/osx/carbon/font.cpp]: yes, sorry for the omission

File: src/osx/carbon/font.cpp (hunk 3 above)
Comment (csomor) [inline src/osx/carbon/font.cpp]: definitely, I'll change that

File: src/osx/carbon/font.cpp (hunk 1 above)
Comment (csomor) [inline src/osx/carbon/font.cpp]: thanks

File: src/osx/carbon/font.cpp [hunk 6]
```diff
@@ -295,6 +295,11 @@ wxFontRefData::wxFontRefData(wxOSXSystemFont font, int size)
 
 static const CGAffineTransform kSlantTransform = CGAffineTransformMake( 1, 0, tan(wxDegToRad(11)), 1, 0, 0 );
 
+typedef struct {
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: This is C-ism, there is really no need for `typedef` in C++, just `struct OSXFontEntry` would do.

Also, not strictly related, but I'd take this inside an anonymous namespace (`namespace {...}`) to prevent this class name from leaking into the global namespace.

File: src/osx/carbon/font.cpp [hunk 7]
```diff
@@ -295,6 +295,11 @@ wxFontRefData::wxFontRefData(wxOSXSystemFont font, int size)
...
+    wxCFRef< CTFontRef > font;
+    wxCFRef< CFDictionaryRef > fontAttributes;
+} OSXFontEntry;
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: A nitpick, but perhaps it would be better to use the word "cache" somewhere in the name here? E.g. `OSXCachedFontEntry` (or even just `CachedFontEntry` because there is no risk of a clash if it's declared inside the anon namespace).

File: src/osx/carbon/font.cpp [hunk 8]
```diff
@@ -313,11 +318,11 @@ void wxFontRefData::MacFindFont()
         // use font caching
         wxString lookupnameWithSize = wxString::Format( "%s_%u_%d", m_info.m_faceName, traits, m_info.m_pointSize );
 
-        static std::map< std::wstring , wxCFRef< CTFontRef > > fontcache ;
-        static std::map< std::wstring , wxCFRef< CFDictionaryRef > > fontattributescache ;
+        static std::map< std::wstring , OSXFontEntry > fontcache ;
```
Comment (vadz) [inline src/osx/carbon/font.cpp]: BTW, I didn't think about it during the original review, but why do we actually need to use `std::wstring` as the map key here? It looks like we could just use `wxString` directly and then we wouldn't even need `ToStdWstring()` call... Of course, this is pretty minor.

**Review by vadz — APPROVED** [review]
Thanks for the updates!

If you commit this yourself, could you please squash all the commits into a single one (it's not really useful to have the individual ones here)? Of course, if you prefer, please let me know and I can do it myself instead.

Thanks again!

**Comment by csomor** [comment]
Thanks a lot Vadim for your reviews, I still have to learn the new ways on how to do things with git and on github: Am I right that this _single commit from pull request_ is done by applying the _Squash and merge_ command from the _Merge pull request_ popup ? If you don't see another optimisation I'd commit then.

**Comment by vslavik** [comment]
Thanks, looks good to me too.  Wouldn't it be better to just use `NSFont` to implement `wxFont` though? There's an awful lot of back-and-forth juggling in `wxFont` and relatively expensive `NSFont` creation that is still the case after these changes (unless I misunderstood them?)...

**Comment by vslavik** [comment]
> is done by applying the Squash and merge command from the Merge pull request popup ?

Yes.

**Comment by csomor** [comment]
Hi Vaclav, thanks very much for your review, and a very good point, I really forgot about these, actually we could remove NSFont/UIFont creation entirely, since macOS 10.5 and iOS 7 CTFontRef is toll-free bridged to NSFont / UIFont. Would it be better to make this into the same commit or into a different pull request ?

**Comment by vadz** [comment]
Please make it a separate commit (it could be the same PR if you want, but then you'd need to squash the existing commits manually instead of using the GitHub button -- it is not difficult to do with `git rebase -i`, but maybe you don't want to bother with this right now...). TIA!
